lightning/src/ln/channelmanager.rs

```diff
@@ -2059,10 +2128,46 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		} else if has_err {
 			Err(PaymentSendFailure::AllFailedRetrySafe(results.drain(..).map(|r| r.unwrap_err()).collect()))
 		} else {
-			Ok(())
+			Ok(payment_id)
 		}
 	}
 
+	/// Retries a payment along the given `Route`.
+	///
+	/// Errors returned are a superset of those returned from `send_payment`, so see `send_payment`
+	/// documentation for more details on errors. This method will also error if the retry amount puts
+	/// the payment nontrivially over the payment's total amount, or if the payment for the given
+	/// `payment_id` cannot be found.
+	pub fn retry_payment(&self, route: &Route, payment_id: PaymentId) -> Result<(), PaymentSendFailure> {
+		let outbounds = self.pending_outbound_payments.lock().unwrap();
+		let (total_msat, payment_hash, payment_secret) = if let Some(payment) = outbounds.get(&payment_id) {
+			match payment {
+				PendingOutboundPayment::Retryable {
+					total_msat, payment_hash, payment_secret, pending_amt_msat, ..
+				} => {
+					let retry_amt_msat: u64 = route.paths.iter().map(|path| path.last().unwrap().fee_msat).sum();
+					if retry_amt_msat as f64 + *pending_amt_msat as f64 > *total_msat as f64 * 1.1 {
+						return Err(PaymentSendFailure::ParameterError(APIError::APIMisuseError {
+							err: format!("retry_amt_msat of {} will put pending_amt_msat (currently: {}) more than 10% over total_payment_amt_msat of {}", retry_amt_msat, pending_amt_msat, total_msat).to_string()
+						}))
+					}
+					(*total_msat, *payment_hash, *payment_secret)
+				},
+				PendingOutboundPayment::Legacy { .. } => {
+					return Err(PaymentSendFailure::ParameterError(APIError::APIMisuseError {
+						err: "Unable to retry payments that were serialized as legacy payments".to_string()
```

Comment:
Maybe s/serialized as legacy payments/initially sent on LDK versions prior to 0.0.102/?